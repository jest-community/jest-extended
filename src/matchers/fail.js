export function fail(_, message) {
  return {
    pass: false,
    message: () => (message ? message : 'fails by .fail() assertion'),
  };
}
