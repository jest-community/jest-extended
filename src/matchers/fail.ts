export function fail(_: unknown, message: string) {
  return {
    pass: false,
    message: () => (message ? message : 'fails by .fail() assertion'),
  };
}
