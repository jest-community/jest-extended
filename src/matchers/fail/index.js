import predicate from './predicate';

const failMessage = message => {
  if (message) return () => message;
  else return () => 'fails by .fail() assertion';
};

export default {
  fail: (expected, message) => ({ pass: predicate(), message: failMessage(message) }),
};
