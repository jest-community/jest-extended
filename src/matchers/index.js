import toBeFalse from './toBeFalse';
import toBeTrue from './toBeTrue';

export default [toBeFalse, toBeTrue].reduce((acc, matcher) => ({ ...acc, ...matcher }), {});
