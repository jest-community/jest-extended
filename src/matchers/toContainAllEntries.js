import { containsEntry } from '../utils';

export function toContainAllEntries(actual, expected) {
  const { printReceived, printExpected, matcherHint } = this.utils;

  const passMessage =
    matcherHint('.not.toContainAllEntries') +
    '\n\n' +
    'Expected object to not only contain all of the given entries:\n' +
    `  ${printExpected(expected)}\n` +
    'Received:\n' +
    `  ${printReceived(actual)}`;

  const failMessage =
    matcherHint('.toContainAllEntries') +
    '\n\n' +
    'Expected object to only contain all of the given entries:\n' +
    `  ${printExpected(expected)}\n` +
    'Received:\n' +
    `  ${printReceived(actual)}`;

  const pass =
    actual.hasOwnProperty &&
    expected.length == Object.keys(actual).length &&
    expected.every(entry => containsEntry(this.equals, actual, entry));

  return { pass, message: () => (pass ? passMessage : failMessage) };
}
