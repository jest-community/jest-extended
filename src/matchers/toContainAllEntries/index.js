import { matcherHint, printExpected, printReceived } from 'jest-matcher-utils';

import predicate from './predicate';

const passMessage = (actual, expected) => () =>
  matcherHint('.not.toContainAllEntries') +
  '\n\n' +
  'Expected object to not only contain all of the given entries:\n' +
  `  ${printExpected(expected)}\n` +
  'Received:\n' +
  `  ${printReceived(actual)}`;

const failMessage = (actual, expected) => () =>
  matcherHint('.toContainAllEntries') +
  '\n\n' +
  'Expected object to only contain all of the given entries:\n' +
  `  ${printExpected(expected)}\n` +
  'Received:\n' +
  `  ${printReceived(actual)}`;

export default {
  toContainAllEntries: (actual, expected) => {
    const pass = predicate(actual, expected);
    if (pass) {
      return { pass: true, message: passMessage(actual, expected) };
    }

    return { pass: false, message: failMessage(actual, expected) };
  }
};
