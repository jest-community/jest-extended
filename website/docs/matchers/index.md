---
sidebar_position: 1
---

# Matchers

- [.pass(message)](/docs/matchers/pass)
- [.fail(message)](/docs/matchers/fail)
- [.toBeEmpty()](/docs/matchers/tobeempty)
- [.toBeOneOf([members])](/docs/matchers/tobeoneof)
- [.toBeNil()](/docs/matchers/tobenil)
- [.toSatisfy(predicate)](/docs/matchers/tosatisfy)

## [Array](/docs/matchers/array)

- [.toBeArray()](/docs/matchers/array/#tobearray)
- [.toBeArrayOfSize()](/docs/matchers/array/#tobearrayofsize)
- [.toIncludeAllMembers([members])](/docs/matchers/array/#toincludeallmembersmembers)
- [.toIncludeAllPartialMembers([members])](/docs/matchers/array/#toincludeallpartialmembersmembers)
- [.toIncludeAnyMembers([members])](/docs/matchers/array/#toincludeanymembersmembers)
- [.toIncludeSameMembers([members])](/docs/matchers/array/#toincludesamemembersmembers)
- [.toPartiallyContain(member)](/docs/matchers/array/#topartiallycontainmember)
- [.toSatisfyAll(predicate)](/docs/matchers/array/#tosatisfyallpredicate)
- [.toSatisfyAny(predicate)](/docs/matchers/array/#tosatisfyanypredicate)
- [.toBeInRange(min, max)](/docs/matchers/array/#tobeinrangemin-max)

## [Boolean](/docs/matchers/boolean)

- [.toBeBoolean()](/docs/matchers/boolean/#tobeboolean)
- [.toBeTrue()](/docs/matchers/boolean/#tobetrue)
- [.toBeFalse()](/docs/matchers/boolean/#tobefalse)

## [Date](/docs/matchers/date)

- [.toBeDate()](/docs/matchers/date/#tobedate)
- [.toBeValidDate()](/docs/matchers/date/#tobevaliddate)
- [.toBeAfter(date)](/docs/matchers/date/#tobeafterdate)
- [.toBeBefore(date)](/docs/matchers/date/#tobebeforedate)
- [.toBeAfterOrEqualTo(date)](/docs/matchers/date/#tobeafterorequaltodate)
- [.toBeBeforeOrEqualTo(date)](/docs/matchers/date/#tobebeforeorequaltodate)
- [.toBeBetween(startDate, endDate)](/docs/matchers/date/#tobebetweenstartdate-enddate)

## [Function](/docs/matchers/function)

- [.toBeFunction()](/docs/matchers/function/#tobefunction)
- [.toThrowWithMessage()](/docs/matchers/function/#tothrowwithmessagetype-message)

## [Mock](/docs/matchers/mock)

- [.toHaveBeenCalledBefore()](/docs/matchers/mock/#tohavebeencalledbefore)
- [.toHaveBeenCalledAfter()](/docs/matchers/mock/#tohavebeencalledafter)
- [.toHaveBeenCalledOnce()](/docs/matchers/mock/#tohavebeencalledonce)
- [.toHaveBeenCalledOnceWith()](/docs/matchers/mock/#tohavebeencalledoncewith)

## [Number](/docs/matchers/number)

- [.toBeNumber()](/docs/matchers/number/#tobenumber)
- [.toBeNaN()](/docs/matchers/number/#tobenan)
- [.toBeFinite()](/docs/matchers/number/#tobefinite)
- [.toBePositive()](/docs/matchers/number/#tobepositive)
- [.toBeNegative()](/docs/matchers/number/#tobenegative)
- [.toBeEven()](/docs/matchers/number/#tobeeven)
- [.toBeOdd()](/docs/matchers/number/#tobeodd)
- [.toBeWithin(start, end)](/docs/matchers/number/#tobewithinstart-end)
- [.toBeInteger()](/docs/matchers/number/#tobeinteger)

## [Object](/docs/matchers/object)

- [.toBeObject()](/docs/matchers/object/#tobeobject)
- [.toBeEmptyObject()](/docs/matchers/object/#tobeemptyobject)
- [.toContainKey(key)](/docs/matchers/object/#tocontainkeykey)
- [.toContainKeys([keys])](/docs/matchers/object/#tocontainkeyskeys)
- [.toContainAllKeys([keys])](/docs/matchers/object/#tocontainallkeyskeys)
- [.toContainAnyKeys([keys])](/docs/matchers/object/#tocontainanykeyskeys)
- [.toContainValue(value)](/docs/matchers/object/#tocontainvaluevalue)
- [.toContainValues([values])](/docs/matchers/object/#tocontainvaluesvalues)
- [.toContainAllValues([values])](/docs/matchers/object/#tocontainallvaluesvalues)
- [.toContainAnyValues([values])](/docs/matchers/object/#tocontainanyvaluesvalues)
- [.toContainEntry([key, value])](/docs/matchers/object/#tocontainentrykey-value)
- [.toContainEntries([[key, value]])](/docs/matchers/object/#tocontainentrieskey-value)
- [.toContainAllEntries([[key, value]])](/docs/matchers/object/#tocontainallentrieskey-value)
- [.toContainAnyEntries([[key, value]])](/docs/matchers/object/#tocontainanyentrieskey-value)
- [.toBeExtensible()](/docs/matchers/object/#tobeextensible)
- [.toBeFrozen()](/docs/matchers/object/#tobefrozen)
- [.toBeSealed()](/docs/matchers/object/#tobesealed)

## [Promise](/docs/matchers/promise)

- [.toResolve()](/docs/matchers/promise/#toresolve)
- [.toReject()](/docs/matchers/promise/#toreject)

## [String](/docs/matchers/string)

- [.toBeString()](/docs/matchers/string/#tobestring)
- [.toBeHexadecimal(string)](/docs/matchers/string/#tobehexadecimal)
- [.toBeDateString(string)](/docs/matchers/string/#tobedatestringstring)
- [.toEqualCaseInsensitive(string)](/docs/matchers/string/#toequalcaseinsensitivestring)
- [.toStartWith(prefix)](/docs/matchers/string/#tostartwithprefix)
- [.toEndWith(suffix)](/docs/matchers/string/#toendwithsuffix)
- [.toInclude(substring)](/docs/matchers/string/#toincludesubstring)
- [.toIncludeRepeated(substring, times)](/docs/matchers/string/#toincluderepeatedsubstring-times)
- [.toIncludeMultiple([substring])](/docs/matchers/string/#toincludemultiplesubstring)
- [.toEqualIgnoringWhitespace(string)](/docs/matchers/string/#toequalignoringwhitespacestring)

## [Symbol](/docs/matchers/symbol)

- [.toBeSymbol()](/docs/matchers/symbol/#tobesymbol)
