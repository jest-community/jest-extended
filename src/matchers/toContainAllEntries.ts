import { containsEntry } from 'src/utils';

export function toContainAllEntries<E = unknown>(
  actual: unknown,
  expected: readonly (readonly [keyof E, E[keyof E]])[],
): {
  pass: boolean;
  message: () => string;
} {
  // @ts-expect-error OK to have implicit any for this
  const { printReceived, printExpected, matcherHint } = this.utils;

  const pass =
    typeof actual === 'object' &&
    actual !== null &&
    !Array.isArray(actual) &&
    expected.length == Object.keys(actual as Record<string, unknown>).length &&
    // @ts-expect-error containsEntry takes an any type
    expected.every(entry => containsEntry(this.equals, actual, entry as [any, any]));

  return {
    pass,
    message: () =>
      pass
        ? matcherHint('.not.toContainAllEntries') +
          '\n\n' +
          'Expected object to not only contain all of the given entries:\n' +
          `  ${printExpected(expected)}\n` +
          'Received:\n' +
          `  ${printReceived(actual)}`
        : matcherHint('.toContainAllEntries') +
          '\n\n' +
          'Expected object to only contain all of the given entries:\n' +
          `  ${printExpected(expected)}\n` +
          'Received:\n' +
          `  ${printReceived(actual)}`,
  };
}
