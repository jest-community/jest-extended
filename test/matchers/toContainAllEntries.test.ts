import * as matcher from 'src/matchers/toContainAllEntries';

expect.extend(matcher);

const data = { a: 'foo', b: 'bar', c: 'baz' };

describe('.toContainAllEntries', () => {
  test('passes when object only contains all of the given entries', () => {
    expect(data).toContainAllEntries([
      ['a', 'foo'],
      ['b', 'bar'],
      ['c', 'baz'],
    ]);
  });

  test('fails when object does not only contain all of the given entries', () => {
    expect(() =>
      expect(data).toContainAllEntries([
        ['a', 'foo'],
        ['b', 'bar'],
      ]),
    ).toThrowErrorMatchingSnapshot();
  });

  test('fails when actual is not an object', () => {
    expect(() =>
      expect(null).toContainAllEntries([
        ['a', 'foo'],
        ['b', 'bar'],
      ]),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      expect(42).toContainAllEntries([
        ['a', 'foo'],
        ['b', 'bar'],
      ]),
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('.not.toContainAllEntries', () => {
  test('passes when object does not only contain all of the given entries', () => {
    expect(data).not.toContainAllEntries([
      ['a', 'qux'],
      ['b', 'bar'],
      ['c', 'baz'],
    ]);
  });

  test('fails when object only contains all of the given entries', () => {
    expect(() =>
      expect(data).not.toContainAllEntries([
        ['b', 'bar'],
        ['a', 'foo'],
        ['c', 'baz'],
      ]),
    ).toThrowErrorMatchingSnapshot();
  });

  test('passes when actual is not an object', () => {
    expect(() =>
      expect(null).not.toContainAllEntries([
        ['a', 'foo'],
        ['b', 'bar'],
      ]),
    );

    expect(() =>
      expect(42).not.toContainAllEntries([
        ['a', 'foo'],
        ['b', 'bar'],
      ]),
    );
  });
});

// Note - custom equality tester must be at the end of the file because once we add it, it cannot be removed
(expect.addEqualityTesters ? describe : describe.skip)('toContainAllEntries with custom equality tester', () => {
  let mockEqualityTester: jest.Mock;
  beforeAll(() => {
    mockEqualityTester = jest.fn();
    expect.addEqualityTesters([mockEqualityTester]);
  });
  afterEach(() => {
    mockEqualityTester.mockReset();
  });
  test('passes when custom equality matches one of the values', () => {
    mockEqualityTester.mockImplementation((a, b) => (a === 'baz' && b === 'bla' ? true : undefined));
    expect(data).toContainAllEntries([
      ['b', 'bar'],
      ['a', 'foo'],
      ['c', 'bla'],
    ]);
  });
  test('fails when custom equality returns false on one of the values', () => {
    mockEqualityTester.mockImplementation((a, b) => (a === 'baz' && b === 'baz' ? false : undefined));
    const entries = Object.entries(data);
    expect(() => expect(data).toContainAllEntries(entries)).toThrowErrorMatchingSnapshot();
  });
});
