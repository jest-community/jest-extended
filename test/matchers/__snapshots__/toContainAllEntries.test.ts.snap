// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.not.toContainAllEntries fails when object only contains all of the given entries 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).not.toContainAllEntries(</intensity><green>expected</color><dim>)</intensity>

Expected object to not only contain all of the given entries:
  <green>[["b", "bar"], ["a", "foo"], ["c", "baz"]]</color>
Received:
  <red>{"a": "foo", "b": "bar", "c": "baz"}</color>"
`;

exports[`.toContainAllEntries fails when actual is not an object 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).toContainAllEntries(</intensity><green>expected</color><dim>)</intensity>

Expected object to only contain all of the given entries:
  <green>[["a", "foo"], ["b", "bar"]]</color>
Received:
  <red>null</color>"
`;

exports[`.toContainAllEntries fails when actual is not an object 2`] = `
"<dim>expect(</intensity><red>received</color><dim>).toContainAllEntries(</intensity><green>expected</color><dim>)</intensity>

Expected object to only contain all of the given entries:
  <green>[["a", "foo"], ["b", "bar"]]</color>
Received:
  <red>42</color>"
`;

exports[`.toContainAllEntries fails when object does not only contain all of the given entries 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).toContainAllEntries(</intensity><green>expected</color><dim>)</intensity>

Expected object to only contain all of the given entries:
  <green>[["a", "foo"], ["b", "bar"]]</color>
Received:
  <red>{"a": "foo", "b": "bar", "c": "baz"}</color>"
`;

exports[`toContainAllEntries with custom equality tester fails when custom equality returns false on one of the values 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).toContainAllEntries(</intensity><green>expected</color><dim>)</intensity>

Expected object to only contain all of the given entries:
  <green>[["a", "foo"], ["b", "bar"], ["c", "baz"]]</color>
Received:
  <red>{"a": "foo", "b": "bar", "c": "baz"}</color>"
`;
